design/003-interactive-session/003-interactive-session.md

```diff
@@ -0,0 +1,123 @@
+[docker](https://www.docker.com/). Docker is one of the most popular container engines, that eases the development
+and deployment of applications, by off-loading the installation of the runtime environment dependencies of the
+application.
+
+In order to start a local interactive session, just as in case of renkulab.io, it is required to provide
+a `Dockerfile`. This is by default provided by `renku init` when creating a renku project, but if the user
+is using a custom project template it might not be present.
+
+### Proposed Commands
+
+#### renku session start
+
+A command to start an interactive session.
+
+By default, the `Dockerfile` present in the given project is going to be used for building the docker image
+that's going to be used for running the interactive session. In case the docker image is not available the
+user is going to be prompted whether the user wants to build the docker image, as in some cases it can be
+time consuming.
+
+With the `--provider <provider_type>` flag a user can specify session provider to be used for starting
+an interactive session. By default two different providers are shipped with Renku CLI:
+ - docker
+ - renkulab
+The provider specific configuration values can be specified by using the `--config <config.yaml>` flag.
+
+##### Detailed Parameter Description
+
+```
+renku session start [OPTIONS]
+
+--image <image_name>			Override the docker image to be used for the interactive session.
+-p|--provider [docker|renkulab] 	Use the specified backend for starting an interactive session.
+-c|--config   				YAML file containing configuration for the provider.
+--cpu <num of cpus>			Specify how much of the available CPU resources a container can use.
+--disk <disk space>			Specify the amount of disk space for the container.
+--gpu <device>				Specify the GPU allocated for the container.
+--memory <max memory>			The maximum amount of memory the container can use.
+
+```
+
+#### renku session stop
+
+`renku session stop my_session` simply stops the given interactive session.
+
+One can shut down all the currently running interactive sessions at once with the `renku session stop --all`.
+
+##### Detailed Parameter Description
+
+```
+renku session stop [OPTIONS] <name>
+
+<name>      Name of the session that the user wants to stop.
+
+--all       Stops all the running containers.
+
+```
+
+#### renku session list
+
+The command to list all the currently running interactive sessions for the given project, with the
+detailed information regarding the network port mappings.
+
+##### Detailed Parameter Description
+
+```
+renku session list [OPTIONS]
+
+-p|--provider [docker|renkulab] 	Use the specified backend for listing the interactive sessions.
+-c|--config   				YAML file containing configuration for the provider.
+
+```
+
+#### renku session open
+
+`renku session open <my_session>` opens the given interactive session in the user's default browser.
```

Comment:
don't we need a `provider` here as well? Or is the provider going to be encoded in the session id returned by `session list`?

Reply:
no, coz the session id is unique so if there's such a session then it'll open it.

Reply:
But you fetch session IDs from the provider - are you going to check all providers for the id?

Reply:
see https://github.com/SwissDataScienceCenter/renku-python/pull/2529/files#diff-9eec9990e65005652022b90320c5717aabc1ea27e3fbdb25441e9bca96a09bc9R64
i.e. it doesn't have `firstresult=True` defined so by default all plugins are going to be executed. so indeed it does go through over all providers. note it's highly unlikely that there's gonna be more than 3-4 providers at once, and this seemed to be a good tradeoff between having the user to be forced to specify a `provider` vs not (and then have all the providers checked).

Reply:
and i dont imagine having like more than a dozen active sessions altogether.
`640kb should be enough for everyone` (c)